Jenkinsfile

```diff
@@ -1,169 +1,158 @@
+      // Make sure not to delete the folder that Jenkins allocates to store scripts
+      sh 'git clean -ffdx -e .????????'
+
+      stage 'Docker build'
+      def buildEnv = docker.build 'realm-java:snapshot'
+      buildEnv.inside("--privileged -v /dev/bus/usb:/dev/bus/usb -v ${env.HOME}/gradle-cache:/root/.gradle -v /root/adbkeys:/root/.android") {
+        stage 'JVM tests'
+        try {
+          gradle 'assemble check javadoc'
+        } finally {
+          storeJunitResults 'realm/realm-annotations-processor/build/test-results/TEST-*.xml'
+          storeJunitResults 'examples/unitTestExample/build/test-results/**/TEST-*.xml'
+          step([$class: 'LintPublisher'])
         }
 
-        currentBuild.rawBuild.setResult(Result.SUCCESS)
-    } catch (Exception e) {
-        echo "The job failed with the following exception: ${e.getMessage()}"
-        currentBuild.rawBuild.setResult(Result.FAILURE)
-    } finally {
-        if (isPullRequest()) {
-            node {
-                reportResultToGithub()
-            }
+        stage 'Static code analysis'
+        try {
+          gradle('realm', 'findbugs pmd checkstyle')
+        } finally {
+          publishHTML(target: [allowMissing: false, alwaysLinkToLastBuild: false, keepAll: true, reportDir: 'realm/realm-library/build/findbugs', reportFiles: 'findbugs-output.html', reportName: 'Findbugs issues'])
+          publishHTML(target: [allowMissing: false, alwaysLinkToLastBuild: false, keepAll: true, reportDir: 'realm/realm-library/build/reports/pmd', reportFiles: 'pmd.html', reportName: 'PMD Issues'])
+          step([$class: 'CheckStylePublisher',
+          canComputeNew: false,
+          defaultEncoding: '',
+          healthy: '',
+          pattern: 'realm/realm-library/build/reports/checkstyle/checkstyle.xml',
+          unHealthy: ''
+          ])
         }
-    }
-}
 
-def String startLogCatCollector() {
-    sh '''adb logcat -c
-    adb logcat > "logcat.txt" &
-    echo $! > pid
-    '''
-    return readFile("pid").trim()
-}
+        stage 'Run instrumented tests'
+        boolean archiveLog = true
+        String backgroundPid
+        try {
+          backgroundPid = startLogCatCollector()
+          gradle('realm', 'connectedUnitTests')
+          archiveLog = false;
+        } finally {
+          stopLogCatCollector(backgroundPid, archiveLog)
+          storeJunitResults 'realm/realm-library/build/outputs/androidTest-results/connected/TEST-*.xml'
+        }
 
-def stopLogCatCollector(String backgroundPid, boolean archiveLog) {
-    sh "kill ${backgroundPid}"
-    if (archiveLog) {
-        zip([
-            'zipFile': 'logcat.zip',
-            'archive': true,
-            'glob' : 'logcat.txt'
+        // TODO: add support for running monkey on the example apps
+
+        if (env.BRANCH_NAME == 'master') {
+          stage 'Collect metrics'
+          collectAarMetrics()
+
+          stage 'Publish to OJO'
+          gradle 'assemble ojoUpload'
```

Comment:
In Groovy if you only have one argument you can omit the brackets and it's used all over the place in the Jenkinsfile DSL as you can see